### PR TITLE
Addiction System Base Code

### PR DESCRIPTION
**Pull request by M11gamer**
The new system used for addictions.
Addiction quirks now work.
This is a base code, that's why this code is barebones and only has two starter addiction quirks (Smoker and Alcoholic).
This code also adds the ability to any reagent to make someone addicted, for now only: Nicotine and Ethanol (Alcohol) can make you addicted.
This may need balancing! Should be tested!

If any code errors or bad code is present please enlighten me, I started working with byond not too long ago.

### CODE REVIEW

File: maps/oldfare/jobs/soldiers.dm [hunk 1]
```diff
@@ -266,4 +266,11 @@ GLOBAL_LIST_INIT(lone_thoughts, list(
 		return
 	var/datum/quirk/chosen_quirk = pick(random_quirks)
 	src.quirk = new chosen_quirk
-	to_chat(src, "<span class='bnotice'>I was formed a bit different. I am [quirk.name]. [quirk.description]</span>")
\ No newline at end of file
+	to_chat(src, "<span class='bnotice'>I was formed a bit different. I am [quirk.name]. [quirk.description]</span>")
+	switch(chosen_quirk)
+		if(/datum/quirk/cig_addict)
```
Comment (mattmatterson111) [inline maps/oldfare/jobs/soldiers.dm:271]: Drop the switch and just make these explicit istype checks isntead.
```suggestion
		if(istype(chosen_quirk, /datum/quirk/cig_addict))
```

File: code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Medicine.dm [hunk 2]
```diff
@@ -278,6 +278,7 @@
 
 /datum/reagent/tramadol/morphine
 	name = "Morphine"
+	reagent_addiction_strength = 15
```
Comment (mattmatterson111) [inline code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Medicine.dm]: While realistic, everyone already takes morphine in this game a lot. People are already basically addicted to it. I would leave addiction for mood boosters like cigarettes and booze, not drugs that are required to function.

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 3]
```diff
@@ -124,3 +145,65 @@
...
+			if("Nicotine")
+				var/message = pick("My mouth is a bit dry...", "A cigarret sure would go down well rigth now")
+				to_chat(M, "<span class='notice'>[message]</span>")
+	return
+
+/datum/reagent/proc/addiction_act_stage2(var/mob/living/carbon/M, var/datum/reagent/R)
+	var/name = check_name(R.name)
+	clear_events(R.name, M)
+	if(prob(10))
+		var/message = pick("I could use some [name] rigth now!","I am feeling a bit stressed, sure could use some [name]!","I could use some [name] rigth now!")
```
Comment (mattmatterson111) [inline code/modules/reagents/Chemistry-Reagents.dm]:
```suggestion
		var/message = pick("I could use some [name] right now!","I am feeling a bit stressed, sure could use some [name]!","I could use some [name] right now!")
```

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 4]
```diff
@@ -124,3 +145,65 @@
 
 /datum/reagent/proc/reaction_mob(var/mob/target)
 	touch_mob(target)
+// Addiction code.... For events
+
+/datum/reagent/proc/addiction_act_stage1(var/mob/living/carbon/M, var/datum/reagent/R)
+	if(prob(5))
+		switch(R.name)
+			if("Nicotine")
+				var/message = pick("My mouth is a bit dry...", "A cigarret sure would go down well rigth now")
```
Comment (mattmatterson111) [inline code/modules/reagents/Chemistry-Reagents.dm]:
```suggestion
				var/message = pick("My mouth is a bit dry...", "A cigarette sure would go down well right now")
```

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 5]
```diff
@@ -124,3 +145,65 @@
...
+	if(prob(10))
+		var/message = pick("I could use some [name] rigth now!","I am feeling a bit stressed, sure could use some [name]!","I could use some [name] rigth now!")
+		to_chat(M, "<span class='notice'>[message]</span>")
+	M.add_event("addiction", /datum/happiness_event/addiction/withdrawal_small)
+	return
+/datum/reagent/proc/addiction_act_stage3(var/mob/living/carbon/M, var/datum/reagent/R)
+	var/name = check_name(R.name)
+	clear_events(R.name, M)
+	if(prob(10))
+		var/message = pick("I really could use some [name] rigth now!","I am feeling a bit stressed, sure could use some [name]!","I should use some [name]")
```
Comment (mattmatterson111) [inline code/modules/reagents/Chemistry-Reagents.dm]:
```suggestion
		var/message = pick("I really could use some [name] right now!","I am feeling a bit stressed, sure could use some [name]!","I should use some [name]")
```

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 6]
```diff
@@ -124,3 +145,65 @@
...
+		to_chat(M, "<span class='notice'>[message]</span>")
+	M.add_event("addiction", /datum/happiness_event/addiction/withdrawal_medium)
+	return
+
+/datum/reagent/proc/addiction_act_stage4(var/mob/living/carbon/M, var/datum/reagent/R)
+	var/name = check_name(R.name)
+	clear_events(R.name, M)
+	if(prob(10))
+		var/message = pick("I really should use some [name] rigth now!","I am stressed, I need to use some [name]!","I should use some [name]")
```
Comment (mattmatterson111) [inline code/modules/reagents/Chemistry-Reagents.dm]:
```suggestion
		var/message = pick("I really should use some [name] right now!","I am stressed, I need to use some [name]!","I should use some [name]")
```

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 7]
```diff
@@ -124,3 +145,65 @@
...
+		to_chat(M, "<span class='notice'>[message]</span>")
+	M.add_event("addiction", /datum/happiness_event/addiction/withdrawal_large)
+	return
+/datum/reagent/proc/addiction_act_stage5(var/mob/living/carbon/M, var/datum/reagent/R)
+	var/name = check_name(R.name)
+	clear_events(R.name, M)
+	if(prob(10))
+		var/message = pick("I need to use some [name] rigth now!","I am stressed, I need to use some [name] RIGTH NOW!","I REALLY NEED SOME [name]")
```
Comment (mattmatterson111) [inline code/modules/reagents/Chemistry-Reagents.dm]:
```suggestion
		var/message = pick("I need to use some [name] right now!","I am stressed, I need to use some [name] RIGHT NOW!","I REALLY NEED SOME [name]")
```

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 8]
```diff
@@ -51,6 +57,21 @@
 		var/overdose_threshold = overdose * (flags & IGNORE_MOB_SIZE? 1 : MOB_MEDIUM/M.mob_size)
 		if(volume > overdose_threshold)
 			overdose(M, alien)
+	if(prob(reagent_addiction_strength) && !is_type_in_list(src, M.reagents.addiction_list))
+		to_chat(M, "<span class='danger'>You like that feeling, you may want more later...</span>")
+		var/datum/reagent/new_reagent = new type()
+		new_reagent.last_addiction_dose = world.timeofday
+		M.reagents.addiction_list.Add(new_reagent)
+	else if(is_type_in_list(src, M.reagents.addiction_list))
+		var/message = pick("I feel better, but for how long?", "Ah.....", "I really needed that")
```
Comment (mattmatterson111) [inline code/modules/reagents/Chemistry-Reagents.dm]:
```suggestion
		var/message = pick("You feel better, but for how long?", "Ah.....")
```

**Review by mattmatterson111 — CHANGES_REQUESTED** [review]

File: maps/oldfare/jobs/soldiers.dm (hunk 1 above)
Comment (M11gamer) [reply maps/oldfare/jobs/soldiers.dm:271]: For some reason, istype does not work in this case, in the start I made it like that but the check would not be true even with the /datum/quirk/cig_addict hardcoded in chosen quirk. @mattroks101

Comment (mattmatterson111) [reply maps/oldfare/jobs/soldiers.dm:271]: Oh? So this method does work for sure then?

Comment (M11gamer) [reply maps/oldfare/jobs/soldiers.dm:271]: For as much as I could test yes. I even put some admin logs to see and only after I changed it from istype to a normal if, I would receive the admin log.

Comment (M11gamer) [reply maps/oldfare/jobs/soldiers.dm:271]: I will test it again.. one moment

Comment (M11gamer) [reply maps/oldfare/jobs/soldiers.dm:271]: It works

File: code/modules/reagents/Chemistry-Holder.dm [hunk 9]
```diff
@@ -4,6 +4,7 @@ GLOBAL_DATUM_INIT(temp_reagents_holder, /obj, new)
 	var/list/datum/reagent/reagent_list = list()
 	var/total_volume = 0
 	var/maximum_volume = 120
+	var/list/datum/reagent/addiction_list = new/list()
```
Comment (mattmatterson111) [inline code/modules/reagents/Chemistry-Holder.dm]:
```suggestion
	var/list/datum/reagent/addiction_list = list()
```

**Review by mattmatterson111 — CHANGES_REQUESTED** [review]

**Review by mattmatterson111 — APPROVED** [review]

**Pull request merged**